### PR TITLE
New gantt tab

### LINKED ISSUE #19940: Create a new gantt chart in React
The Gantt chart could become far more useful to see how a dag is performing. We should make it a first class citizen by integrating it with the tree view.

The gantt rows can be lined up with the tree view to keep context.
It would be easy to explore different runs by selecting separate columns in the tree view.

Without a dag run selected we can have a summary across many dag runs. Even including a box and whisker plot to show the median run times and min/max.

If you have selected an active run, the gantt view should also auto-refresh.

### PR DESCRIPTION
**Pull request by bbovenzi**
New Gantt Chart in grid new tabs

Closes: [#19940](https://github.com/apache/airflow/issues/19940)

Sync scrolling with grid:
Jun-08-2023 11-47-07

Support task groups and sync with grid:
Jun-08-2023 17-11-01

Sync task/run selection with grid:
Jun-08-2023 17-12-47

To Do:
- Improve scroll/height calculation. 
- Fix and add tests

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
Whoa. I don't have skills to review it but the screenshots looks like #protm

**Comment by eladkal** [comment]
this looks really good!
What happens when task retry and it will show the timeline only for last attempt right?
I wonder if it's doable to allow users choose to see all retries (as requested in https://github.com/apache/airflow/issues/17487 )

**Comment by marclamberti** [comment]
my eyes 🤩🤩🤩🤩🤩

**Comment by jens-scheffler-bosch** [comment]
WOW and only 440 LoC! This smells like PR of the month is coming :-D

**Comment by bbovenzi** [comment]
Ready for review although I'm still working on improving the scroll behavior.

File: airflow/www/static/js/dag/details/gantt/Chart.tsx [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+              borderBottomWidth={1}
+              borderBottomColor={isGroup && isOpen ? "gray.400" : "gray.200"}
+              bg={isSelected ? "blue.100" : "inherit"}
+            >
+              {instance ? (
+                <Tooltip
+                  label={
+                    <Box>
+                      <Text>
+                        Task{isGroup ? " Group" : ""}: {task.label}
```
Comment (pierrejeambrun) [inline airflow/www/static/js/dag/details/gantt/Chart.tsx]: Can we extract a component from this. I find it hard to read to have all this code inlined in the jsx.

File: airflow/www/static/js/dag/details/index.tsx [hunk 2]
```diff
@@ -262,6 +295,16 @@ const Details = ({ openGroupIds, onToggleGroups, hoveredTaskState }: Props) => {
               hoveredTaskState={hoveredTaskState}
             />
           </TabPanel>
+          {run && (
+            <TabPanel p={0} height="100%">
+              <Gantt
+                openGroupIds={openGroupIds}
+                // hoveredTaskState={hoveredTaskState}
```
Comment (pierrejeambrun) [inline airflow/www/static/js/dag/details/index.tsx]: We can remove if unnecessary, tasks are not hovering in the Gantt tab, maybe 'nice to have' but not necessary IMHO

File: airflow/www/static/js/dag/details/index.tsx [hunk 3]
```diff
@@ -134,6 +153,12 @@ const Details = ({ openGroupIds, onToggleGroups, hoveredTaskState }: Props) => {
     [setSearchParams, searchParams, showLogs, showMappedTasks, taskId]
   );
 
+  useEffect(() => {
+    if ((!taskId || isGroup) && tabIndex > 3) {
+      onChangeTab(1);
+    }
+  }, [runId, taskId, tabIndex, isGroup, onChangeTab]);
```
Comment (pierrejeambrun) [inline airflow/www/static/js/dag/details/index.tsx]: Is `runId` necessary here ?

**Review by pierrejeambrun — COMMENTED** [review]
Looks great! Tested locally and seems nice. :tada: 

I just have issues scrolling the gantt. (scroll bars are not working for me), also I have vertical scroll when there is plenty of space vertically to render the Gantt chart. (scrollbar are unresponsive, both with the scroll wheel and by click/drag on it, tested on chrome and firefox on ubuntu)

image

File: airflow/www/static/js/dag/details/index.tsx (hunk 3 above)
Comment (bbovenzi) [reply airflow/www/static/js/dag/details/index.tsx]: Updated. There was an existing bug with resetting the tabs.

File: airflow/www/static/js/dag/details/index.tsx (hunk 2 above)
Comment (bbovenzi) [reply airflow/www/static/js/dag/details/index.tsx]: Removed

File: airflow/www/static/js/dag/details/gantt/Chart.tsx (hunk 1 above)
Comment (bbovenzi) [reply airflow/www/static/js/dag/details/gantt/Chart.tsx]: Cleaned up!

**Comment by bbovenzi** [comment]
> Looks great! Tested locally and seems nice. 🎉
> 
> I just have issues scrolling the gantt. (scroll bars are not working for me), also I have vertical scroll when there is plenty of space vertically to render the Gantt chart. (scrollbar are unresponsive, both with the scroll wheel and by click/drag on it, tested on chrome and firefox on ubuntu)

Figured out the scroll focus issue on the gantt chart itself. I also change the `checkScrollPosition` from `onSelect` to a timeout, so if the scroll of the grid/gantt are not synced, it should rectify itself.

**Review by ashb — APPROVED** [review]
LGTM. In this case I think we could even remove the original Gantt chart entirely. Happy to make that a separate PR if you want. (Either way, when we delete that we should redirect the old URL to the new react-based one)

**Comment by eladkal** [comment]
>  In this case I think we could even remove the original Gantt chart entirely

+1

**Comment by potiuk** [comment]
> In this case I think we could even remove the original Gantt chart entirely

+1

File: airflow/www/static/js/dag/details/gantt/GanttTooltip.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        <Time dateTime={instance?.startDate} />
+      </Text>
+      <Text>
+        {instance?.endDate ? (
+          <>
+            {isMappedOrGroupSummary && "Latest "}End:{" "}
+            <Time dateTime={instance?.endDate} />
+          </>
+        ) : (
+          "Ongoing"
```
Comment (jedcunningham) [inline airflow/www/static/js/dag/details/gantt/GanttTooltip.tsx]: We should consider having nothing instead. This makes for a weird combo for queued tasks:
Screenshot 2023-07-26 at 4 46 05 PM

**Review by pierrejeambrun — APPROVED** [review]

**Pull request merged**